shapely/ops.py

```diff
@@ -553,14 +553,62 @@ def split(geom, splitter):
 
 
 def substring(geom, start_dist, end_dist, normalized=False):
-    """Return a line segment between specified distances along a linear geometry.
+    """Return a line segment between specified distances along a LineString
 
     Negative distance values are taken as measured in the reverse
     direction from the end of the geometry. Out-of-range index
     values are handled by clamping them to the valid range of values.
-    If the start distances equals the end distance, a point is being returned.
-    If the normalized arg is True, the distance will be interpreted as a
-    fraction of the geometry's length.
+
+    If the start distance equals the end distance, a Point is returned.
+
+    If the start distance is actually beyond the end distance, then the
+    reversed substring is returned such that the start distance is
+    at the first coordinate.
+
+    Parameters
+    ----------
+    geom : LineString
+        The geometry to get a substring of.
+    start_dist : float
+        The distance along `goem` of the start of the substring.
+    end_dist : float
+        The distance along `geom` of the end of the substring.
+    normalized : bool, False
+        Whether the distance parameters are interpreted as a
+        fraction of the geometry's length.
+
+    Returns
+    -------
+    Union[Point, LineString]
+        The substring between `start_dist` and `end_dist` or a Point
+        if they are at the same location.
+
+    Raises
+    ------
+    AssertionError
+        If `geom` is not a LineString.
+
+    Examples
+    --------
+    >>> from shapely.geometry import LineString
+    >>> from shapely.ops import substring
+    >>> ls = LineString((i, 0) for i in range(6))
+    >>> ls.wkt
+    'LINESTRING (0 0, 1 0, 2 0, 3 0, 4 0, 5 0)'
+    >>> substring(ls, start_dist=1, end_dist=3).wkt
+    'LINESTRING (1 0, 2 0, 3 0)'
+    >>> substring(ls, start_dist=3, end_dist=1).wkt
+    'LINESTRING (3 0, 2 0, 1 0)'
+    >>> substring(ls, start_dist=1, end_dist=-3).wkt
+    'LINESTRING (1 0, 2 0)'
+    >>> substring(ls, start_dist=0.2, end_dist=-0.6, normalized=True).wkt
+    'LINESTRING (1 0, 2 0)'
+
+    Returning a `Point` when `start_dist` and `end_dist` are at the
+    same location.
+
+    >>> substring(ls, 2.5, -2.5).wkt
+    'POINT (2.5 0)'
     """
 
     assert(isinstance(geom, LineString))
```

Comment:
Hmm... tabbing through `raise TypeError` and `raise ValueError` in the module, it looks like `transform`, `nearest_points`, and `split` all use `ValueError` for bad types while only `shared_paths` uses `TypeError`. Unless this misses a majority which are raised by sub-functions, it looks like the `ops` module tends to use `ValueError`. 

If I were starting from scratch, I think I would prefer `TypeError` because it seems to be what the language suggests. But I can also see a useful pattern where you just catch `ValueError` for mistakes that you make with your arguments. Anyway, I'm not sure a discussion about that preference would matter in light of the consistency issue. 

Does the information about which error is more common in `ops` change your suggestion?

Reply:
given the inconsistency at the moment (see #1035), just pick one for now and we'll adjust it later.